crates/p2p_protocol/protocol/p2p.capnp

```diff
@@ -0,0 +1,179 @@
+# lib3h p2p wire protocol
+# note, initial handshake will use transit-enc.capnp
+# once that is complete, messages from this schema
+# will be encoded as negotiated.
+
+@0x859198991b95d3e1;
+
+struct Message {
+  # Main P2P Message struct, anon union determines message type
+
+  union {
+    msgError @0 :MsgError;
+    # indicates an error
+    # if the error is not recoverable, the connection will be closed
+
+    msgHandshake @1 :MsgHandshake;
+    # On a new connection, tell the remote node about ourselves.
+
+    msgGspArcRequest @2 :MsgGspArc;
+    # Open a gossip sequence with a remote node.
+
+    msgGspArcResponse @3 :MsgGspArc;
+    # Second stage of a gossip sequence with a remote node.
+
+    msgGspAspectDataRequest @4 :MsgGspAspectDataRequest;
+    # If we have determined a list of aspect-hashes we require
+    # from a remote node, request them.
+
+    msgGspAspectDataResponse @5 :MsgGspAspectDataResponse;
+    # A remote node has requested aspect data from us, give it to them.
+
+    msgGspAspectBroadcast @6 :MsgGspAspectBroadcast;
+    # Fast push new dht data
+
+    msgDirectRequest @7 :MsgDirect;
+    # node-to-node message request
+
+    msgDirectResponse @8 :MsgDirect;
+    # node-to-node message response
+
+    msgQueryRequest @9 :MsgQuery;
+    # dht query request
+
+    msgQueryResponse @10 :MsgQuery;
+    # dht query response
+  }
+
+  # -- top-level Message Types -- #
+
+  struct MsgError {
+    errorCode @0 :ErrorCode;
+    # code indicating if error is well-known
+
+    errorText @1 :Text;
+    # text indicating details of error
+
+    enum ErrorCode {
+      unknown @0;
+      # default if error is not well-known, or if remote is using a newer proto
+    }
+  }
+
+  struct MsgHandshake {
+    fromStoreArc @0 :UInt32;
+    # the storage arc length of the source transport / agent
+
+    fromQueryArc @1 :UInt32;
+    # the query arc length of the source transport / agent
+  }
+
+  struct MsgGspArc {
+    # data for `msgGspArcRequest` and `msgGspArcResponse`
+
+    aspectConstraintLoc @0 :AspectConstraintLoc;
+    # constrain aspectHashList by loc arc
+
+    aspectConstraintTime @1 :AspectConstraintTime;
+    # constrain aspectHashList by publish timestamp
+
+    aspectConstraintCount @2 :AspectConstraintCount;
+    # constrain aspectHashList by local store count
+
+    aspectHashList @3 :List(AspectHashList);
+    # list of aspectHashes associated with entryAddresses
+    # that fall within all the above constraints.
+  }
```

Comment:
Suggest renaming `aspectHashList` to `aspectHashListResponse` or something of this sort to make clear that the field is used for the response only.

Reply:
Actually this is used in both directions. The node making the request sends every aspectHash within the constraints, then the responding node sends only those hashes that both fall within the constraints AND the requesting node didn't include in the request. (because they already have those).

Reply:
Then both nodes individually request data for all the hashes they don't already know about.

Reply:
Oh okay, sweet! Forget my initial comment then.